// cf-nodejs-logging-support/build/main/lib/middleware/framework-services/fastify.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class FastifyService {
    getReqHeaderField(req, fieldName) {
        return req.headers[fieldName];
    }
    getReqField(req, fieldName) {
        var _a, _b, _c;
        let value = undefined;
        switch (fieldName) {
            case "protocol":
                value = "HTTP" + (req.raw.httpVersion == null ? "" : "/" + req.raw.httpVersion);
                break;
            case "remote_host":
                value = (_a = req.raw.connection) === null || _a === void 0 ? void 0 : _a.remoteAddress;
                break;
            case "remote_port":
                value = (_c = (_b = req.raw.connection) === null || _b === void 0 ? void 0 : _b.remotePort) === null || _c === void 0 ? void 0 : _c.toString();
                break;
            case "remote_user":
                if (req.raw.user && req.raw.user.id) {
                    value = req.raw.user.id;
                }
                break;
            default:
                value = req.raw[fieldName];
                break;
        }
        return value;
    }
    getResHeaderField(res, fieldName) {
        return res.getHeader ? res.getHeader(fieldName) : undefined;
    }
    getResField(res, fieldName) {
        return res.raw[fieldName];
    }
    onResFinish(res, handler) {
        res.raw.on("header", handler);
        res.raw.on("finish", handler);
    }
}
exports.default = FastifyService;
